### PR TITLE
Feature/radius ldap better integration

### PR DESCRIPTION
**Pull request by fdurand**
# Description

* New service pfcertmanager that generate certificate files based on the configuration.
* EAP configuration template, add a way to define multiples eap profiles in freeradius.
* Choose which eap profile to use based on a radius filter.
* Basedn can be define in the authentication sources rules
* New hooks for the radius filter engine in eduroam virtual server.
* Redefined "restart" in the service manager to allow "PartOf" in systemcd scripts
* Set role from source authentication rule option (needs #5459)
* Flatten the radius request for the authentication sources (attributes like radius_request.User-Name)
* Radius request attributes / Username are part of the common attributes.
* Support of multiples ldap servers in frreeradius ldap_packetfence configuration file
* copy outter user-Name attribute in PacketFence-Outer-User attribute to be able to use it in the authentication rules.
* copy the LDAP-UserDN attribute in PacketFence-UserDN attribute to be able to use it in the authentication rules.
* Added a way to extend the ldap filter for searchattributes configuration

# Impacts
* RADIUS authorize workflow
* LDAP Authentication sources.

# Code / PR Dependencies

PR #5459

# NEW Package(s) required
freeradius >= 3.0.21

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* New service pfcertmanager that generate certificate files based on the configuration.
* EAP configuration template, add a way to define multiples eap profiles in freeradius.

## Enhancements
* Choose which eap profile to use based on a radius filter.
* Basedn can be define in the authentication sources rules
* New hooks for the radius filter engine in eduroam virtual server.
* Redefined "restart" in the service manager to allow "PartOf" in systemcd scripts
* Set role from source authentication rule option (needs #5459)
* Flatten the radius request for the authentication sources (attributes like radius_request.User-Name)
* Radius request attributes / Username are part of the common attributes.
* Support of multiples ldap servers in frreeradius ldap_packetfence configuration file
* copy outter user-Name attribute in PacketFence-Outer-User attribute to be able to use it in the authentication rules.
* copy the LDAP-UserDN attribute in PacketFence-UserDN attribute to be able to use it in the authentication rules.
* Added a way to extend the ldap filter for searchattributes configuration

### CODE REVIEW

**Comment by satkunas** [comment]
`not_deletable` needs to be added to the search endpoints for all sections in order to supress the DELETE button on the table lists.

`/usr/local/pf/sbin/pfperl-api get -M GET "/api/v1/config/radiusd/tls_profiles?cursor=0&limit=25&fields=id,certificate_profile,ocsp&sort=id" | jq`
```json
{
  "items": [
    {
      "certificate_profile": "radius",
      "id": "test",
      "ocsp": "default"
    },
    {
      "certificate_profile": "radius",
      "id": "tls-common",
      "ocsp": "default"
    },
    {
      "certificate_profile": "radius",
      "id": "tls-eap-fast",
      "ocsp": "default"
    },
    {
      "certificate_profile": "radius",
      "id": "tls-eap-fast-copy",
      "ocsp": "default"
    }
  ],
  "prevCursor": 0,
  "status": 200,
  "total_count": 4
}
```

**Comment by jrouzierinverse** [comment]
> `not_deletable` needs to be added to the search endpoints for all sections in order to supress the DELETE button on the table lists.
> 
> `/usr/local/pf/sbin/pfperl-api get -M GET "/api/v1/config/radiusd/tls_profiles?cursor=0&limit=25&fields=id,certificate_profile,ocsp&sort=id" | jq`
> 
> ```json
> {
>   "items": [
>     {
>       "certificate_profile": "radius",
>       "id": "test",
>       "ocsp": "default"
>     },
>     {
>       "certificate_profile": "radius",
>       "id": "tls-common",
>       "ocsp": "default"
>     },
>     {
>       "certificate_profile": "radius",
>       "id": "tls-eap-fast",
>       "ocsp": "default"
>     },
>     {
>       "certificate_profile": "radius",
>       "id": "tls-eap-fast-copy",
>       "ocsp": "default"
>     }
>   ],
>   "prevCursor": 0,
>   "status": 200,
>   "total_count": 4
> }
> ```
Ask and you shall receive.
```bash
/usr/local/pf/sbin/pfperl-api get -M GET "/api/v1/config/radiusd/tls_profiles?cursor=0&limit=25&fields=id,certificate_profile,ocsp,not_deletable&sort=id" | jq

{
  "items": [
    {
      "certificate_profile": "radius",
      "id": "tls-common",
      "not_deletable": true,
      "ocsp": "default"
    },
    {
      "certificate_profile": "radius",
      "id": "tls-eap-fast",
      "not_deletable": true,
      "ocsp": "default"
    }
  ],
  "prevCursor": 0,
  "status": 200,
  "total_count": 2
}
```

**Comment by fdurand** [comment]
@jrouzierinverse and @satkunas i will need to have inheritance for those files :

```
conf/radiusd/eap_profiles.conf.defaults
conf/radiusd/fast.conf.defaults
conf/radiusd/ocsp.conf.defaults
conf/radiusd/tls.conf.defaults
```

So all the defaults profiles (eap/fast/ocsp/tls) can be editable and override the values we have in the .default files.

The only one who don't inherit and is not editable and is hidden is the conf/ssl.conf.defaults

**Comment by julsemaan** [comment]
We need upgrade scripts to migrate the deprecated parameters of pf.conf into what I believe is a new TLS configuration

File: go/certmanager/main.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+			ctx := ctx
+			go func(ctx context.Context) {
+				for {
+					pfconfigdriver.PfconfigPool.Refresh(ctx)
+					time.Sleep(1 * time.Second)
+				}
+			}(ctx)
+		})
+	} else {
+		panic("Unable to obtain pfconfigpool lock in certmanager")
```
Comment (julsemaan) [inline go/certmanager/main.go]: Not sure you should panic here, logging the error should be enough

File: lib/pf/services/manager.pm [hunk 2]
```diff
@@ -498,6 +488,50 @@ sub sysdDisable {
...
+
+=cut
+
+sub restartService {
+    my ($self) = @_;
+    my $cmdLine = $self->restart_launcher;
+    if ($cmdLine =~ /^(.+)$/) {
+        $cmdLine = $1;
+        my $logger = get_logger();
+        $logger->debug(sprintf("Restartarting Daemon %s with command %s",$self->name,$cmdLine));
```
Comment (julsemaan) [inline lib/pf/services/manager.pm]: I think its Restartartartartarting

**Comment by julsemaan** [comment]
We will also need doc on how to make simple EAP-TLS work with this new mechanism. No need to explain the super complex workflow we can create but just instructions on how to make the current simple case we support work with this

File: lib/pf/services/manager.pm (hunk 2 above)
Comment (fdurand) [reply lib/pf/services/manager.pm]: It's when it's really hard to start.

File: go/certmanager/main.go (hunk 1 above)
Comment (fdurand) [reply go/certmanager/main.go]: To be honest i did a copy and paste from the function in pfdns.
I will log instead of panic.

**Comment by fdurand** [comment]
> We will also need doc on how to make simple EAP-TLS work with this new mechanism. No need to explain the super complex workflow we can create but just instructions on how to make the current simple case we support work with this

I will rework this part and we will be able to select a EAP config from the realm configuration. It will be easier to configure/understand.

**Comment by julsemaan** [comment]
Holding on for the final review for:

- [x] Rebase of the PR
- [x] Upgrade script for EAP-TLS that still needs to be done

**Pull request merged**